cluster/kubernetes/scoper.go

```diff
@@ -1,33 +1,112 @@
 package kubernetes
 
 import (
+	"math/rand"
+	"sync"
+	"time"
+
 	"k8s.io/apimachinery/pkg/api/errors"
-	"k8s.io/client-go/discovery"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/weaveworks/flux/cluster"
 )
 
+var (
```

Comment:
Would this suffice? https://github.com/kubernetes/client-go/blob/master/discovery/cached/memcache.go (does not use memcached!)

Reply:
I considered it and discarded it because:

1.  It caches to disk (see `~/.kube/cache` in the environment you normally use for `kubectl`):
       * We only need the `namespaced` bit of every resource, which makes it overkill to store to disk given the small memory requirements
       * Reading from memory is much quicker anyhow
2. It doesn't support TTLs (at least explicitly), which I think are needed in case the resource definitions change (e.g. CRDs being changed or added incrementally).

Reply:
I don't see the bit where it writes to disk (in the code in that file). Is that in the broader caching system?

Reply:
Ah, you are right. I missed that, sorry. I had just looked at https://github.com/kubernetes/client-go/blob/master/discovery/cached_discovery.go which I had previously looked at, and thought you were referring to that (the file layout is a bit odd).

I guess we can periodically invalidate the whole cache with `Invalidate()` to make up for the lack of explicit TTLs. However, the cache implementation is a bit aggressive, since it's all or nothing (it gets the information of all groups at once) resulting in ~50 requests every time (based on the number of version groups from my local k8s cluster).

Reply:
The newer sync implementation in #1442 relies pretty heavily on the discovery interface -- it uses it to probe for the kinds of resources that the cluster knows about, so it can ask for all of the resources. In anticipation of that, I am less concerned about caching the whole set of endpoints.

Invalidation is trickier to judge. CRDs could change namespacedness, as you say (though this is a niche problem, getting something wrong, like incorrectly deleting a CRD and that cascading to all the custom resources, would be bad). On the other hand, even synchronously requesting all the metadata would not be immune to problems of that nature -- it needs to be solved some other way, if it can be.

Reply:
If CRDs are the only thing we think could change, we could watch for CRD changes and invalidate on those.